consensus/system_contract/consensus.go

```diff
@@ -0,0 +1,407 @@
+	}
+	// Ensure that the difficulty is one
+	if header.Difficulty.Cmp(common.Big1) != 0 {
+		return errInvalidDifficulty
+	}
+	// Verify that the gas limit is <= 2^63-1
+	cap := uint64(0x7fffffffffffffff)
+	if header.GasLimit > cap {
+		return fmt.Errorf("invalid gasLimit: have %v, max %v", header.GasLimit, cap)
+	}
+	// All basic checks passed, verify cascading fields
+	return s.verifyCascadingFields(chain, header, parents)
+}
+
+// verifyCascadingFields verifies all the header fields that are not standalone,
+// rather depend on a batch of previous headers. The caller may optionally pass
+// in a batch of parents (ascending order) to avoid looking those up from the
+// database. This is useful for concurrently verifying a batch of new headers.
+func (s *SystemContract) verifyCascadingFields(chain consensus.ChainHeaderReader, header *types.Header, parents []*types.Header) error {
+	// The genesis block is the always valid dead-end
+	number := header.Number.Uint64()
+	if number == 0 {
+		return nil
+	}
+	// Ensure that the block's timestamp isn't too close to its parent
+	var parent *types.Header
+	if len(parents) > 0 {
+		parent = parents[len(parents)-1]
+	} else {
+		parent = chain.GetHeader(header.ParentHash, number-1)
+	}
+	if parent == nil || parent.Number.Uint64() != number-1 || parent.Hash() != header.ParentHash {
+		return consensus.ErrUnknownAncestor
+	}
+	if header.Time < parent.Time {
+		return errInvalidTimestamp
+	}
+	// Verify that the gasUsed is <= gasLimit
+	if header.GasUsed > header.GasLimit {
+		return fmt.Errorf("invalid gasUsed: have %d, gasLimit %d", header.GasUsed, header.GasLimit)
+	}
+	if !chain.Config().IsCurie(header.Number) {
+		// Verify BaseFee not present before EIP-1559 fork.
+		if header.BaseFee != nil {
+			return fmt.Errorf("invalid baseFee before fork: have %d, want <nil>", header.BaseFee)
+		}
+		if err := misc.VerifyGaslimit(parent.GasLimit, header.GasLimit); err != nil {
+			return err
+		}
+	} else if err := misc.VerifyEip1559Header(chain.Config(), parent, header); err != nil {
+		// Verify the header's EIP-1559 attributes.
+		return err
+	}
+
+	if err := s.correctSigner(header); err != nil {
+		return err
+	}
+	return nil
+}
+
+// correctSigner checks that the actual block signer matches
+// the L1-authorized signer for this block's number.
+func (s *SystemContract) correctSigner(header *types.Header) error {
+	actualSigner, err := ecrecover(header)
+	if err != nil {
+		return err
+	}
+
+	expectedSigner, err := s.findSignerSlice(header.Number.Uint64())
+	if err != nil {
+		return err // e.g. "no signer for block"
+	}
+
+	if actualSigner != expectedSigner {
+		return errUnauthorizedSigner
+	}
+	return nil
+}
+
+func (s *SystemContract) findSignerSlice(blockNum uint64) (common.Address, error) {
+	// Acquire read-lock before accessing s.signerAddressesL1
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	transitions := s.signerAddressesL1
+	idx := sort.Search(len(transitions), func(i int) bool {
```

Comment:
to me this looks like an overkill and a too comlicated for what it needs to do. if we have a DESC sorted list based on startBlock, then we can just iterate through it and if `blockNum > startBlock` we found the signer. the list should be short enough and in most executions (the node is synced and live) it will find the correct singer in the first loop iteration.

Reply:
Good point, I will make the list DESC sorted when inserting (l1 contract PR). Thanks.

Reply:
Update: no need to do this as we are going for now with one single signer.